plugins/basic/python/tk_maya_basic/plugin_logic.py

```diff
@@ -0,0 +1,207 @@
+# agreement to the Shotgun Pipeline Toolkit Source Code License. All rights
+# not expressly granted therein are reserved by Shotgun Software Inc.
+
+import logging
+
+import pymel.core as pm
+
+# For now, import the Shotgun toolkit core included with the plug-in,
+# but also re-import it later to ensure usage of a swapped in version.
+import sgtk
+
+from sgtk_plugin_basic import manifest
+import plugin_engine
+import plugin_logging
+
+from . import __name__ as PLUGIN_PACKAGE_NAME
+
+
+MENU_LOGIN = "ShotgunMenuLogin"
+MENU_LABEL = "Shotgun"
+
+ITEM_LABEL_LOGIN   = "Log In to Shotgun..."
+ITEM_LABEL_LOGOUT  = "Log Out of Shotgun"
+ITEM_LABEL_WEBSITE = "Learn about Shotgun..."
+
+WEBSITE_URL = "https://shotgunsoftware.com"
+
+
+# Initialize a standalone logger to display messages in Maya script editor
+# when the engine is not running while the user is logged out of Shotgun.
+standalone_logger = logging.getLogger(PLUGIN_PACKAGE_NAME)
+# Do not propagate messages to Maya ancestor logger to avoid duplicated logs.
+standalone_logger.propagate = False
+# Ignore messages less severe than the debug ones.
+standalone_logger.setLevel(logging.DEBUG)
+# Use a custom logging handler to display messages in Maya script editor.
+standalone_logger.addHandler(plugin_logging.PluginLoggingHandler(manifest.name))
+
+
+def bootstrap():
+    """
+    Bootstraps the plug-in logic handling user login and logout.
+    """
+
+    if sgtk.authentication.ShotgunAuthenticator().get_default_user():
+        # When the user is already authenticated, automatically log him/her in.
+        login_user()
+    else:
+        # When the user is not yet authenticated, display a login menu.
+        create_login_menu()
+
+
+def shutdown():
+    """
+    Shutdowns the plug-in logic handling user login and logout.
+    """
+
+    if sgtk.platform.current_engine():
+        # When the user is logged in with a running engine, shut down the engine.
+        plugin_engine.shutdown()
+    else:
+        # When the user is logged out, delete the displayed login menu.
+        delete_login_menu()
+
+
+def login_user():
+    """
+    Logs in the user to Shotgun and starts the engine.
+    """
+
+    try:
+        # When the user is not yet authenticated,
+        # pop up the Shotgun login dialog to get the user's credentials,
+        # otherwise, get the cached user's credentials.
+        user = sgtk.authentication.ShotgunAuthenticator().get_user()
+
+    except sgtk.authentication.AuthenticationCancelled:
+        # When the user cancelled the Shotgun login dialog,
+        # keep around the displayed login menu.
+        standalone_logger.info("Shotgun login was cancelled by the user.")
+        return
+
+    # Get rid of the displayed login menu since the engine menu will take over.
+    delete_login_menu()
+
+    # Before bootstrapping the engine for the first time around,
+    # the toolkit manager may swap the toolkit core to its latest version.
+    plugin_engine.bootstrap(user,
```

Comment:
I wonder if this method has the right name. It's calling the bootstrap, not just logging the user in.